WifiDirectTest/app/src/main/java/test/microsoft/com/wifidirecttest/WifiDirectService.java

```diff
@@ -0,0 +1,411 @@
+    static final public String DSS_WIFIDIRECT_MESSAGE = "test.microsoft.com.wifidirecttest.DSS_WIFIDIRECT_MESSAGE";
+
+
+    private static final String TAG = MainActivity.class.getPackage().getName();
+
+    private static final String SERVICE_TYPE = "_p2p._tcp";
+
+    LocalBroadcastManager broadcaster;
+    MyTextSpeech mySpeech = null;
+
+    private WifiP2pManager p2p;
+    private WifiP2pManager.Channel channel;
+    private volatile ServerSocket wifiSocket = null;
+
+    //can we make this dynamic ?
+    private static final int PORT = 45678;
+
+    private BroadcastReceiver receiver;
+    private IntentFilter filter;
+
+    private WifiP2pManager.PeerListListener peerListListener;
+    private WifiP2pManager.DnsSdServiceResponseListener serviceListener;
+    private WifiP2pManager.DnsSdTxtRecordListener txtListener;
+
+    private WifiP2pDnsSdServiceInfo serviceInfo;
+
+    private final IBinder mBinder = new LocalBinder();
+    public class LocalBinder extends Binder {
+        WifiDirectService getService() {
+            return WifiDirectService.this;
+        }
+    }
+    @Override
+    public IBinder onBind(Intent intent) {
+        return mBinder;
+    }
+
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        return START_STICKY;
+    }
+
+    @Override
+    public void onCreate() {
+        mySpeech = new MyTextSpeech(this);
+
+        broadcaster = LocalBroadcastManager.getInstance(this);
+
+
+        p2p = (WifiP2pManager) getSystemService(WIFI_P2P_SERVICE);
+        if (p2p == null) {
+            mySpeech.speak("This device does not support Wi-Fi Direct");
+            sendResult("This device does not support Wi-Fi Direct");
+        } else {
+            channel = p2p.initialize(this, getMainLooper(), null);
+
+            receiver = new PeerReceiver();
+            filter = new IntentFilter();
+            filter.addAction(WIFI_P2P_STATE_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_PEERS_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_CONNECTION_CHANGED_ACTION);
+            filter.addAction(WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
+            registerReceiver(receiver, filter);
+
+            openWifiSocket();
+
+            peerListListener = new WifiP2pManager.PeerListListener() {
+
+                public void onPeersAvailable(WifiP2pDeviceList peers) {
+                    sendResult("Discovered peers:");
+
+                    final WifiP2pDeviceList pers = peers;
+                    String allInOne = "";
+                    int numm = 0;
+                    for (WifiP2pDevice peer : pers.getDeviceList()) {
+                        numm++;
+                        allInOne = allInOne + MyP2PHelper.deviceToString(peer) + ", ";
+                        sendResult("\t" + MyP2PHelper.deviceToString(peer));
+                    }
+                    mySpeech.speak(numm + " peers discovered, names: " + allInOne);
```

Comment:
Wouldn't any summary that is worth speaking out also be worth putting into the log? In other words I was expected to see a sendSpokenResult() function that both calls mySpeech.speak and logs the string as well. This will make reading the logs even easier.